### PR TITLE
Implement support for multiple locations in CxxReportSensor

### PR DESCRIPTION
**Pull request by ivangalkin**
enable multiple locations for
* CxxDrMemorySensor
* CppcheckParserV2
* CxxValgrindSensor
enable the new interface for all remaining sensors

this pull request is replacement for #1436
see also issue  #1440

---
This change is [Reviewable](https://reviewable.io/reviews/sonaropencommunity/sonar-cxx/1447)

### CODE REVIEW

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/drmemory/CxxDrMemorySensor.java [hunk 1]
```diff
@@ -82,22 +87,39 @@ protected void processReport(final SensorContext context, File report) {
...
-          break;
-        }
+      if (!locations.isEmpty()) {
+        CxxReportIssue issue = new CxxReportIssue(CxxDrMemoryRuleRepository.KEY, error.getType().getId(), locations);
+        saveUniqueViolation(context, issue);
       }
     }
   }
 
+  private void addFileLocations(final SensorContext context, DrMemoryError error, List<CxxReportLocation> locations) {
```
Comment (ivangalkin) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/drmemory/CxxDrMemorySensor.java]: same logic is implemented in CxxValgrindSensor

For each CxxReportLocation CxxReportSensor will check if path belongs to the project or not. If it doesn't - this [secondary] location will be ignored. So finally we will have 

* a primary location which is guaranteed to belong to the project (same as before)
* ```>= 0``` additional locations, which reference files from the analyzed project;

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportIssue.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+/**
+ * Issue with one or multiple locations
+ */
+public class CxxReportIssue {
```
Comment (ivangalkin) [inline cxx-sensors/src/main/java/org/sonar/cxx/sensors/utils/CxxReportIssue.java]: no logic here - CxxReportIssue and CxxReportLocation (1:n) were introduced more or less in order to implement the proper semantics of ::equals() and ::hashCode() methods

**Comment by ivangalkin** [comment]
@guwirth Thank you for your proposal. I started to implement the `CxxReportIssue` in the way you suggested. However the existing code is very careful w.r.t. the XML (or other source) validation, so we save an issue only if at least 1 valid location exists. This makes the generic logic work like that:

1. inspect and collect all possible locations
2. if there is at least 1 valid one -> save the issue

Creation of an `CxxReportIssue` in advance is often impossible or makes the logic more complex.

Also there is a (minor) performance aspect (or rather memory footprint). Since the most of sensors support only primary location we can benefit from the usage of `Collections.singletonList()`. The suggested approach would always require a full-blown container.

File: cxx-sensors/src/main/java/org/sonar/cxx/sensors/cppcheck/CppcheckParserV2.java [hunk 3]
```diff
@@ -70,18 +76,18 @@ public void stream(SMHierarchicCursor rootCursor) throws XMLStreamException {
...
-                String id = errorCursor.getAttrValue("id");
-                String msg = createMsg(
-                  errorCursor.getAttrValue("inconclusive"),
-                  errorCursor.getAttrValue("msg")
-                );
-                String file = null;
-                String line = null;
+                String id = Objects.requireNonNull(errorCursor.getAttrValue("id"),
+                    "Missing mandatory attribute /results/errors/error[@id]");
+                String msg = Objects.requireNonNull(errorCursor.getAttrValue("msg"),
```
[the remark this answers is not in the document]
Comment (ivangalkin) [reply cxx-sensors/src/main/java/org/sonar/cxx/sensors/cppcheck/CppcheckParserV2.java]: thx! done

**Comment by ivangalkin** [comment]
@Bertk please review manually, GitHub doesn't notice, that I've changed the SonarQube finding. I've fixed an issue, but in an other line as reviewed.

**Comment by guwirth** [comment]
@ivangalkin 

> if there is at least 1 valid one -> save the issue

No location means issue on module/project level.

> Creation of an CxxReportIssue in advance is often impossible or makes the logic more complex.

Maybe ctor without location and only this `addLocation `method? If you need an extra check why not add an `isValid` method to `CxxReportIssue`?

**Comment by ivangalkin** [comment]
@guwirth I'll try to apply the comments in the next few days.

**Pull request merged**